src/schedule_synthesis.jl

```diff
@@ -0,0 +1,273 @@
+for each comprising `Automaton` of the `_SynthesizedAutomaton` to compute a 
+list of individual states. For example
+
+```julia
+julia> _state_separation(6, [1, 2])
+[1, 2]
+```
+The explanation of the example is as follows
+```
+6 -> [1, 1, 0]
+[1, 1, 0] / [1, 2] -> [[1], [1, 0]]
+[[1], [1, 0]] -> [1, 2]
+```
+"""
+function _state_separation(l::Int64, B::Vector{Int64}; indigits=false)
+    @assert l < 2^sum(B) "l has more bits than the sum of B"
+    
+    bits = digits(l, base=2, pad=sum(B)) |> reverse
+    index_points = cumsum(B)
+    from_index = [0; index_points[1:end-1]]
+    to_index = index_points
+    bits_separated = [bits[i+1:j] for (i, j) in zip(from_index, to_index)]
+    if indigits
+        bits_separated
+    else
+        map(_undigit, bits_separated)
+    end
+end
+
+"""
+Struct for a controller automaton which enforces only the weakly 
+hard constraint for that controller. i.e., it is not concerned with the
+dynamical matrix of the system.
+"""
+struct _ConstraintAutomaton
+    # # of locations. Legal locations are in the range 0:L-1.
+    L::Int64
+    # # of actions. Legal actions are in the range 0:Σ-1.
+    Σ::Int64
+    # Transition function. T(l,σ) is a location in 0:L-1.
+    T::Function
+    # Initial location in L.
+    l_0::Int64
+    # Function that returns whether a location is final
+    Q_f::Function
+end
+
+"""
+Struct for a synthesized automaton from multiple constraint automata.
+"""
+struct _SynthesizedAutomaton
+    # # of comprising automata
+    N::Int64
+    # List of bits for all comprising controllers
+    B::Vector{Int64}
+    # Locations. Legal locations are in the range 0:L-1.
+    L::Int64
+    # List of actions.
+    Σ::Vector{Int64}
+    # Transition function.  T(l,σ) is a location in 0:L-1.
+    T::Function
+    # Initial location in L.
+    l_0::Int64
+    # Function that returns whether a location is final.
+    Q_f::Function
+end
+
+"""
+Build an constraint controller automaton for the given weakly hard constraint
+"""
+function _controller_automaton(c::WeaklyHardConstraint)
+    # Define the automaton's structure. State l=0 is the dead state (trapping)
+    if c.meet == 0
+        # No requirement. Always valid.
+        L = 2
+        T = (l, σ) -> 1
+    elseif c.meet == c.window
+        # No misses allowed.
+        L = 2
+        T = (l, σ) -> l & σ
+    else
+        # Build full automaton. Dead state is l = 1
+        L = 2^c.window
+        T = function (l, σ)
+            l_new = (l << 1) & (L - 1) | σ
+            l == 0 || count_ones(l_new) < c.meet ? 0 : l_new
+        end
+    end
+    
+    # Put it all together. Starting position is L-1 since there is no miss from the beginning
+    _ConstraintAutomaton(L, 2, T, L-1, !iszero)
+end
+
+"""
+Build a scheduler automaton from a given list of controller automata
+"""
+function _scheduler_automaton(controllers::Vector{_ConstraintAutomaton}; slotsize::Int64=1)
```

Comment:
This could be called `_SynthesizedAutomaton` to make it clearer what it's an external constructor for.